lib/Test/NoLeaks.pm

```diff
@@ -0,0 +1,337 @@
+allocation for perl process, and the leak will be detected.
+
+Page size is 4kb on linux, so, if C<code> leaks 4 bytes on every
+pass, then C<1024> passes should be specified.
+
+In general, the more passes are specified, the more chance to
+detect possible leaks.
+
+Default value is C<100>. Minimal value is C<2>.
+
+=item * C<warmup_passes>
+
+How many times the C<code> should be executed before module starts
+tracking resources consumption on executing the C<code> C<passes>
+times.
+
+If you have caches, memoizes etc., then C<warmup_passes> is your
+friend.
+
+Default value is C<0>.
+
+=item * C<tolerate_hits>
+
+How many passes, which considered leaked, should be ingnored, i.e.
+maximal number of possible false leak reports.
+
+Even if your code has no leaks, it might cause perl interpreter
+allocate additional memory pages, e.g. due to memory fragmentation.
+Those allocations are legal, and should not be treated as leaks.
+
+Default value is C<0>.
+
+=back
+
+=cut
+
+my $PAGE_SIZE;
+
+BEGIN {
+    no strict "subs";
+    *_platform_mem_size = _linux_mem_size;
+    *_platform_fds = _linux_fds;
+    $PAGE_SIZE = sysconf(_SC_PAGESIZE)
+      or die("page size cannot be determined, Test::NoLeaks cannot be used");
+}
+
+sub _linux_mem_size {
+    my $class = shift;
+
+    open(my $fh, '<', '/proc/self/statm')
+        or die("couldn't access /proc/self/status : $!");
+
+    my $line = <$fh>;
+    my ($pages) = (split / /, $line)[0];
+    return $pages * $PAGE_SIZE;
+}
+
+sub _linux_fds {
+    my $fd_dir = '/proc/self/fd';
+    opendir(my $dh, $fd_dir) || die "can't opendir $fd_dir: $!";
+    my $open_fd_count = () = readdir($dh);
+    closedir $dh;
+    return $open_fd_count;
+}
+
+
+sub _noleaks(%) {
+    my %args = @_;
+
+    # check arguments
+    my $code = $args{code};
+    die("code argument (CODEREF) isn't provided")
+        if (!$code || !(ref($code) eq 'CODE'));
+
+    my $track_memory = $args{'track_memory'};
+    my $track_fds    = $args{'track_fds'};
+    die("don't know what to track (i.e. no 'track_memory' nor 'track_fds' are specified)")
+        if (!$track_memory && !$track_fds);
+
+    my $passes = $args{passes} || 100;
+    die("passes count too small (should be at least 2)")
+        if $passes < 2;
+
+    my $warmup_passes = $args{warmup_passes} || 0;
+    die("warmup_passes count too small (should be non-negative)")
+        if $passes < 0;
+
+    my %leaked_at; # key: pass, value array[$mem_leak, $fds_leak]
+
+    # warm-up phase
+    # a) warm up code
+    $code->() for (1 .. $warmup_passes);
+
+    # b) warm-up package itself, as it might cause additional memory (re) allocations
+    # (ignore results)
+    _platform_mem_size if $track_memory;
+    _platform_fds if $track_fds;
+    %leaked_at = map { $_ => [0, 0] } (1 .. $passes);
+
+    # reset warmed-up leaked_at hash
+    %leaked_at = ();
+
+    my ($total_mem_leak, $total_fds_leak, $memory_hits) = (0, 0, 0);
```

Comment:
better if you initialize all these variables including those in the loop below before you measure the size for the first time. Also, better to initialize `@leaked_at` with all-zeroes before and use something like this:

```perl
$leaked_at[$pass]->[0] += $leaked_mem;
$leaked_at[$pass]->[1] += $leaked_fds;
```